hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlPrunedMemberLeftTest.java

```diff
@@ -0,0 +1,125 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.config.MapConfig;
+import com.hazelcast.config.PartitioningAttributeConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.jet.core.JetTestSupport;
+import com.hazelcast.jet.datamodel.Tuple2;
+import com.hazelcast.jet.sql.impl.misc.Pojo;
+import com.hazelcast.map.IMap;
+import com.hazelcast.spi.impl.NodeEngine;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static com.hazelcast.jet.sql.SqlTestSupport.createMapping;
+import static com.hazelcast.jet.sql.impl.SqlEndToEndTestSupport.calculateExpectedPartitions;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class SqlPrunedMemberLeftTest extends JetTestSupport {
+    // Page size is 4096, so we'll fetch 2 pages.
+    private static final int MAP_ENTRIES = 5_000;
+    private static final int CLUSTER_SIZE = 5;
+    private static final int CONSTANT = 2;
+    private static final String MAP_NAME = "map";
+
+    private HazelcastInstance[] hz;
+    private HazelcastInstance client;
+    private NodeEngine masterNodeEngine;
+    private Map<Address, HazelcastInstance> addrToInstanceMap;
+
+    @Before
+    public void setUp() throws Exception {
+        hz = new HazelcastInstance[CLUSTER_SIZE];
+        addrToInstanceMap = new LinkedHashMap<>();
+        for (int i = 0; i < CLUSTER_SIZE; ++i) {
+            hz[i] = createHazelcastInstance(regularInstanceConfig());
+            addrToInstanceMap.put(hz[i].getCluster().getLocalMember().getAddress(), hz[i]);
+        }
+        client = createHazelcastClient();
+        masterNodeEngine = getNodeEngineImpl(hz[0]);
+
+        prepareMap();
+        createMapping(client, MAP_NAME, Pojo.class, Integer.class);
+    }
+
+    @Test
+    // https://hazelcast.atlassian.net/browse/HZ-2992
+    public void test() {
+        // Given
+        String query = "SELECT this FROM " + MAP_NAME + " WHERE f0 = " + CONSTANT;
+
+        HazelcastInstance prunedInstanceToTerminate = findPrunedInstance();
+
+        int cnt = 0;
+
+        for (SqlRow row : client.getSql().execute(query)) {
+            // Shutdown chosen member before we read anything.
+            if (cnt == 0) {
+                prunedInstanceToTerminate.shutdown();
```

Comment:
this test is unreliable. It passes in ~2/3 cases when I change partitioning column in MapConfig to f1. In such case the test should always fail.

`shutdown` AFAIK is async and you do wait for the instance shutdown before continuing iteration.

Reply:
and even unmodified test is flaky when run in a loop, but sometimes first execution fails

Reply:
Yes, in this iteration the test is flaky, because the client is connecting to a random member. When you terminate the member client connected to, the test fails.

Reply:
I removed the client from this test, because it is impossible(most probably) to control to which member it connects.